### ISSUE TITLE
[Proposal] Ini configuration, model set property changes.

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#1374]
### Multiple inout support is not possible (#681, #1253)
- fetching multiple inputs, labels not possible.
- label is copied and pasted where `requireLabel() == true` but each layer might expect some different labels.
- label shape is assumed to be shape of the output, it is not always true.

### Setting a flexible dataset is not possible
- When setting a file for trainset, it is not possible to set generator for validation or test data.

```ini
[dataset]
train_data='foo'
valid_data='bar'
```

## Proposed solution

Previously, it was discussed in #1253 to put input_layers inside `[dataset]` section, but as in this proposed solution, there are multiple section to describe dataset, it is moved to `[model]` section.

### 1. ini description (equivalent setting will be available in api)
```ini
[model] #describe input and label inside the section
input_layers=Layer1,Layer2,Layer3
label_layers=Layer3

[layer1]
type=fully_connected
input_shape=1:1:10 # when specified as an input_layers, setting input_shape is mandatory

[layer2]
type = input
input_shape=2:3:5

[layers3]
type = conv
input_shape=1:2:3
label_shape=2:3:5 # when specified as a label_layers, setting label_shape is required.
```

### 2. handling edge cases

- setting through `[dataset] section` will be left for backward compatibility.
- when input_layers not given, we assume there is only one input, if more then one input_shape is given without connection, we throw.
- when label_layers not given, we copy the same label to all the layer which `requiresLabel() == true`

**Comment by kparichay** [comment]
This might be unrelated to this issue, but does every layer has to support input_shape and label_shape?
Currently conv and fc layer additionally support input_shape, and this example shows conv supporting label_shape. So, I wonder if that is intentional?

**Comment by kparichay** [comment]
> ```
> [model] #describe input and label inside the section
> input_layers=Layer1,Layer2,Layer3
> label_layers=Layer3
> ```

I recommend to update NeuralNet compile interface as below to match with the proposed design
```diff
-int compile();
+int compile(std::vector<std::string> inputs, std::vector<std::string> outputs);
```

You also might want to think about the renaming being done by the graph to ensure unique names for each layer. If name-based api is to be exposed, then it must be ensured that names mentioned in the ini model are not changed and must remain as given.
Further, some api to find the name of the layer/find layer by name might be good to open.

**Comment by zhoonit** [comment]
> This might be unrelated to this issue, but does this design also propose every layer to support input_shape and label_shape?
> Currently, convolution and fully connected layer additionally support input_shape, and this example shows convolution layer supporting label_shape. So, I wonder if that is intentional?

This is to follow current semantics, in this way input 
The intention here was to show whatever layer will except a label. (but we will check `requireLabel() == true` so above example will fail.)

> As a side topic, I propose to provide label via another input_layer. This way, all the external data will always be fed via two fixed entry points, a list of input layers for inputs and a list of input layers for labels. In this case, the new keywork label_shape is no longer required and input_shape can still do the job. If we do this, InputLayer can be renamed to DataSource layers.

This sounds a good alternative. One concern is that we will have additional input layer for a loss making less convenient. But moving to `label_shape` will be easier to support later as per request, so I'll start with this idea.

**Comment by zhoonit** [comment]
> > ```
> > [model] #describe input and label inside the section
> > input_layers=Layer1,Layer2,Layer3
> > label_layers=Layer3
> > ```
> 
> I recommend to update NeuralNet compile interface as below to match with the proposed design
> 
> ```diff
> -int compile();
> +int compile(std::vector<std::string> inputs, std::vector<std::string> outputs);
> ```
> 
> You also might want to think about the renaming being done by the graph to ensure unique names for each layer. If name-based api is to be exposed, then it must be ensured that names mentioned in the ini model are not changed and must remain as given.
> Further, some api to find the name of the layer/find layer by name might be good to open.

I have one concern that this will deviate from capi. 
Changing signature of the `compile` can be easily done later, so I'll not include this in this issue for now, but keep in mind for a discussion. As per the additional api suggestion, I'll check if we will have usage of such api while making tests :)

**Comment by kparichay** [comment]
The capi will also have to support this. This can be done under variable number of arguments, modification of the API or adding a new function, but must be kept while supporting the changes via ini.

**Comment by zhoonit** [comment]
Yes, ini support will be supported via setProperty methods. The concern is about changing signature of the compile :)

**Comment by kparichay** [comment]
I am not sure if setProperty is the right way to do it, it is not a property but rather some information defining the structure for the graph.

**Comment by kparichay** [comment]
> This sounds a good alternative. One concern is that we will have additional input layer for a loss making less convenient. But moving to `label_shape` will be easier to support later as per request, so I'll start with this idea.

The intention is to limit the flow of data always via input_layer. This enforces a uniform and standard location to describe the shape of data and how the dataset interacts with the model. Further, it also simplifies out the scenario of special handling of memory for two scenarios of label and input, where in fact is just one case of external data.

Currently, all types of loss layers will have to support label_shape and the layers with direct input data support have to support the input_shape property. Further, the graph also has to worry about these things and based on the variety of layers/where it is in the graph (start/end/middle). Unifying the inflow of external data simplifies all of these scenarios.

IMO it might be good to directly start with sourcing data via selected layers, otherwise, you might have to handle all these cases in the implementation.

**Comment by zhoonit** [comment]
> I am not sure if setProperty is the right way to do it, it is not a property but rather some information defining the structure for the graph.

Arguably it is not property. 
But setProperty 1:1 maps to ini. So, it is a property of a model. (similar to keras api). Otherwise it does not 1:1 maps to capi <-> ccapi <-> ini.

**Comment by zhoonit** [comment]
> Currently, all types of loss layers will have to support label_shape and the layers with direct input data support have to support the input_shape property. Further, the graph also has to worry about these things and based on the variety of layers/where it is in the graph (start/end/middle). Unifying the inflow of external data simplifies all of these scenarios.

I totally agree about the point you've made. It is safer for us to only allow 'ports' to receive inputs/labels.
Things we have to consider:
1. receiving input shapes for a common layer is already allowed. https://github.com/Samsung/tizen-docs/blob/master/docs/application/native/guides/machine-learning/machine-learning-training.md#layer
2. for a single loss, single label, overhead (in user side) to create a new input layer.

But for now, as suggested, I'll make only input layer to handle this. and later we can extend as needed.

**Comment by kparichay** [comment]
> 1. receiving input shapes for a common layer is already allowed. https://github.com/Samsung/tizen-docs/blob/master/docs/application/native/guides/machine-learning/machine-learning-training.md#layer

I think this property can be deprecated.

> 2. for a single loss, single label, overhead (in user side) to create a new input layer.

For single layer, user can directly specify loss in model, and IMO it should all automatically work out :).

**Comment by kparichay** [comment]
> > I am not sure if setProperty is the right way to do it, it is not a property but rather some information defining the structure for the graph.
> 
> Arguably it is not property.
> But setProperty 1:1 maps to ini. So, it is a property of a model. (similar to keras api). Otherwise it does not 1:1 maps to capi <-> ccapi <-> ini.

Either is fine I guess, via setProperty() or arguments to compile().

**Comment by zhoonit** [comment]
There are some undone work but 90% of the proposal has been implemented. I have updated the undone work. let's track it separately.
